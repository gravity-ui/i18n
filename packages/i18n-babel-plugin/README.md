# @gravity-ui/i18n-babel-plugin

Babel-плагин для оптимизации поставки файлов с переводами.

Если вы не используете babel, то вам подойдет [i18n-optimize-plugin](../i18n-optimize-plugin/README.md).

*Трансформирует файлы `i18n.ts`:*

- Удаляет meta из сообщений (id, description, etc.)
- Преобразует markdown в html (при `meta.markdown === true`)
- Применяет типограф к содержимому ключей

*TODO:*

- Проверяет синтаксис согласно ICU MessageFormat
- Компилирует ICU MessageFormat в AST
- Заменяет исходные ключи на хэши (позволяет сократить длину ключа)

## Использование

Установите плагин:

```bash
npm install @gravity-ui/i18n-babel-plugin --save-dev
```

Подключите плагин к сборке. Пример, если вы используете `app-builder`:

```ts
import {defineConfig} from '@gravity-ui/app-builder';

export default defineConfig({
    client: {
        babel: (config) => {
            const plugins = config.plugins ? [...config.plugins] : [];

            return {
                ...config,
                plugins: [
                    ...plugins,
                    require.resolve('@gravity-ui/i18n-babel-plugin')
                ],
            };
        }
    }
});
```

## Настройки

### typograf

Позволяет настроить [правила типографа](https://github.com/typograf/typograf/blob/dev/docs/RULES.ru.md).

По-умолчанию использует правила [`DEFAULT_TYPOGRAF_CONFIG`](./src/typograf.ts).

Пример кастомизации правил типографа:

```ts
plugins: [
    require.resolve('@gravity-ui/i18n-babel-plugin', {
        typograf: {
            enabled: ['common/nbsp/afterNumber', 'common/nbsp/afterParagraphMark'],
            disabled: ['common/symbols/cf', 'common/space/trimRight', 'common/space/trimLeft'],
        }
    })
],
```

Пример отключения типографа:

```ts
plugins: [
    require.resolve('@gravity-ui/i18n-babel-plugin', {
        typograf: false,
    })
],
```

### root

Позволяет ограничить область работы плагина. Плагин не будет обрабатывать файлы за пределами заданной области.

Тип: `string`.

По-умолчанию: `src`.
