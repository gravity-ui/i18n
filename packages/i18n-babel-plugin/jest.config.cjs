/** @type {import('ts-jest').JestConfigWithTsJest} */
module.exports = {
    preset: 'ts-jest',
    testEnvironment: 'node',
    testMatch: ['<rootDir>/src/__tests__/*.spec.ts'],
    setupFilesAfterEnv: ['<rootDir>/jest.setup.js'],
};
