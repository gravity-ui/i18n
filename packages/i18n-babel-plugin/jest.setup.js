/* eslint-disable no-undef */
// Mock browser APIs that aren't available in Node.js
global.File = class File {
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    constructor(fileBits, fileName, options) {
        this.name = fileName;
        this.size = 0;
        this.type = options?.type || '';
        this.lastModified = Date.now();
        this.webkitRelativePath = '';
    }

    arrayBuffer() {
        return Promise.resolve(new ArrayBuffer(0));
    }

    text() {
        return Promise.resolve('');
    }

    stream() {
        return new ReadableStream();
    }

    slice() {
        return new File([], this.name, {type: this.type});
    }
};

// Mock FileReader if needed
global.FileReader = class FileReader {
    constructor() {
        this.readyState = 0;
        this.result = null;
        this.error = null;
    }

    addEventListener() {}
    removeEventListener() {}
    dispatchEvent() {}
    abort() {}
    readAsArrayBuffer() {}
    readAsDataURL() {}
    readAsText() {}
};

// Mock other File API related globals if needed
global.FileList = class FileList extends Array {
    item(index) {
        return this[index];
    }
};

global.Blob = class Blob {
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    constructor(array, options) {
        this.size = 0;
        this.type = options?.type || '';
    }

    arrayBuffer() {
        return Promise.resolve(new ArrayBuffer(0));
    }

    text() {
        return Promise.resolve('');
    }

    stream() {
        return new ReadableStream();
    }

    slice() {
        return new Blob([], {type: this.type});
    }
};
