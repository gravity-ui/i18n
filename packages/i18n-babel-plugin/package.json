{
  "name": "@gravity-ui/i18n-babel-plugin",
  "version": "1.0.0",
  "license": "UNLICENSED",
  "main": "./dist/index.js",
  "typings": "./dist/index.d.ts",
  "scripts": {
    "build": "pnpm run build:clean && pnpm run build:compile",
    "build:compile": "tsc -p tsconfig.json",
    "build:clean": "rm -rf dist",
    "lint": "eslint .",
    "lint:fix": "eslint . --fix",
    "test": "jest",
    "typecheck": "tsc --noEmit"
  },
  "dependencies": {
    "@babel/core": "^7.26.0",
    "@babel/helper-plugin-utils": "^7.25.9",
    "@babel/traverse": "^7.26.4",
    "@diplodoc/transform": "^4.63.2",
    "@formatjs/icu-messageformat-parser": "^2.9.4",
    "@gravity-ui/i18n-types": "workspace:*",
    "typograf": "^7.4.0"
  },
  "devDependencies": {
    "@babel/types": "^7.26.3",
    "@types/babel__core": "^7.20.5",
    "@types/babel__helper-plugin-utils": "^7.10.3",
    "@types/babel__traverse": "^7.20.6",
    "@types/estree": "^1.0.6",
    "@types/jest": "^29.5.14",
    "@types/node": "^24.7.0",
    "eslint": "^9.15.0",
    "eslint-config": "workspace:*",
    "jest": "29.7.0",
    "nx": "21.6.3",
    "prettier": "^3.4.2",
    "ts-jest": "29.1.1",
    "tslib": "^2.8.1",
    "tsx": "^4.20.6"
  },
  "engines": {
    "node": ">= 18"
  },
  "nx": {}
}
