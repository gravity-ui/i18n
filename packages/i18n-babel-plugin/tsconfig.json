{
    "extends": "../../tsconfig.base.json",
    "compilerOptions": {
        "outDir": "./dist",
        "target": "ES2021",
        "lib": ["ES2021"],
        "module": "Node16",
        "moduleResolution": "Node16"
    },
    "include": ["src/*.ts", "src/**/*.ts", "src/__tests__/index.spec.tsx"],
    "exclude": ["**/__tests__/**"]
}
