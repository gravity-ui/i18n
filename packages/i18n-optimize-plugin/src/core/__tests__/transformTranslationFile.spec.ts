import {TECH_LOCALE} from '@gravity-ui/i18n-babel-plugin';
import {
    includeDynamicImportToTranslationsFile,
    transformFileWithBabel,
} from '../transformTranslationFile';

const SIMPLE_MESSAGE = `
import {intl} from './intl';

export const {t} = intl.createMessages({
    normalMessage: {
        ru: 'русский',
        en: 'english',
        meta: {
            id: '123456',
        },
    },
})`;

const MARKDOWN_MESSAGE = `
import {intl} from './intl';

export const {t} = intl.createMessages({
    messageWithMarkdown: {
        ru: 'Содержат правила для получения и отправки трафика. Управлять группами безопасности можно в сервисе [Virtual Private Cloud](https://console.yandex.cloud/link/vpc/security-groups){target=_blank rel="noopener"}.\\n\\nГруппы безопасности не ограничивают трафик к веб-интерфейсу Apache Airflow™.',
        en: 'Contain rules for receiving and sending traffic. You can manage security groups in the [Virtual Private Cloud service](https://kz.console.yandex.cloud/link/vpc/security-groups){target=_blank rel="noopener"}.\\n\\nSecurity groups do not restrict traffic to the Apache Airflow™ web interface.',
        meta: {
            markdown: true,
        },
    },
});
`;

const LONG_MESSAGE_FOR_TYPOGRAPH = `
import {intl} from './intl';

export const {t} = intl.createMessages({
    normalMessage: {
        ru: 'русский язык длинное предложение pnpm сомнительно но окей',
        en: 'pnpm is good package manager',
        meta: {
            id: '123456',
        },
    },
})`;

const SIMPLE_MESSAGE_WITH_ALISES = `
import {intl} from './intl';

export const {t: commonT, Message: CommonMessage} = intl.createMessages({
    normalMessage: {
        ru: 'русский',
        en: 'english',
        meta: {
            id: '123456',
        },
    },
})`;

describe('transformFileWithBabel', () => {
    it('remove meta from messages', async () => {
        const result = await transformFileWithBabel(SIMPLE_MESSAGE);
        expect(result.code).toMatchSnapshot();
    });

    it('transform markdown to html', async () => {
        const result = await transformFileWithBabel(MARKDOWN_MESSAGE);
        expect(result.code).toMatchSnapshot();
    });

    it('with typograf', async () => {
        const result = await transformFileWithBabel(LONG_MESSAGE_FOR_TYPOGRAPH);
        expect(result.code).toMatchSnapshot();
    });

    it('without typograf', async () => {
        const result = await transformFileWithBabel(LONG_MESSAGE_FOR_TYPOGRAPH, {
            typograf: false,
        });
        expect(result.code).toMatchSnapshot();
    });

    it('transform to TECH_LOCALE', async () => {
        const result = await transformFileWithBabel(SIMPLE_MESSAGE, {
            mode: 'only-translations',
            allowedLocales: [TECH_LOCALE],
        });
        expect(result.code).toMatchSnapshot();
    });
});

describe('includeDynamicImportToTranslationsFile', () => {
    it('should include dynamic import to translations file', () => {
        const result = includeDynamicImportToTranslationsFile(SIMPLE_MESSAGE);
        expect(result.code).toMatchSnapshot();
    });

    it('should include dynamic import to translations file with aliases', () => {
        const result = includeDynamicImportToTranslationsFile(SIMPLE_MESSAGE_WITH_ALISES);
        expect(result.code).toMatchSnapshot();
    });
});
