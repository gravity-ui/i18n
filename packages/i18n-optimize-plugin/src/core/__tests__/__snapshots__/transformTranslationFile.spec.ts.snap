// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`includeDynamicImportToTranslationsFile should include dynamic import to translations file 1`] = `
"
import {intl} from './intl';

const langs = require.context("./.compiled-locales/", false, /\\/.*\\.ts$/);
const messages = langs("./" + process.env.I18N_LANGUAGE + ".ts").default;
export const {t} = intl.createMessages(messages);
"
`;

exports[`includeDynamicImportToTranslationsFile should include dynamic import to translations file with aliases 1`] = `
"
import {intl} from './intl';

const langs = require.context("./.compiled-locales/", false, /\\/.*\\.ts$/);
const messages = langs("./" + process.env.I18N_LANGUAGE + ".ts").default;
export const {t: commonT, Message: CommonMessage} = intl.createMessages(messages);
"
`;

exports[`transformFileWithBabel remove meta from messages 1`] = `
"import { intl } from './intl';
export const {
  t
} = intl.createMessages({
  normalMessage: {
    "ru": "\\u0440\\u0443\\u0441\\u0441\\u043A\\u0438\\u0439",
    "en": "english"
  }
});"
`;

exports[`transformFileWithBabel transform markdown to html 1`] = `
"import { intl } from './intl';
export const {
  t
} = intl.createMessages({
  messageWithMarkdown: {
    "ru": "<p>\\u0421\\u043E\\u0434\\u0435\\u0440\\u0436\\u0430\\u0442 \\u043F\\u0440\\u0430\\u0432\\u0438\\u043B\\u0430 \\u0434\\u043B\\u044F \\u043F\\u043E\\u043B\\u0443\\u0447\\u0435\\u043D\\u0438\\u044F \\u0438\\xA0\\u043E\\u0442\\u043F\\u0440\\u0430\\u0432\\u043A\\u0438 \\u0442\\u0440\\u0430\\u0444\\u0438\\u043A\\u0430. \\u0423\\u043F\\u0440\\u0430\\u0432\\u043B\\u044F\\u0442\\u044C \\u0433\\u0440\\u0443\\u043F\\u043F\\u0430\\u043C\\u0438 \\u0431\\u0435\\u0437\\u043E\\u043F\\u0430\\u0441\\u043D\\u043E\\u0441\\u0442\\u0438 \\u043C\\u043E\\u0436\\u043D\\u043E \\u0432\\xA0\\u0441\\u0435\\u0440\\u0432\\u0438\\u0441\\u0435 <a href=\\"https://console.yandex.cloud/link/vpc/security-groups\\" target=\\"_blank\\" rel=\\"noopener\\">Virtual Private Cloud</a>.</p>\\n<p>\\u0413\\u0440\\u0443\\u043F\\u043F\\u044B \\u0431\\u0435\\u0437\\u043E\\u043F\\u0430\\u0441\\u043D\\u043E\\u0441\\u0442\\u0438 \\u043D\\u0435\\xA0\\u043E\\u0433\\u0440\\u0430\\u043D\\u0438\\u0447\\u0438\\u0432\\u0430\\u044E\\u0442 \\u0442\\u0440\\u0430\\u0444\\u0438\\u043A \\u043A\\xA0\\u0432\\u0435\\u0431-\\u0438\\u043D\\u0442\\u0435\\u0440\\u0444\\u0435\\u0439\\u0441\\u0443 Apache Airflow\\u2122.</p>\\n",
    "en": "<p>Contain rules for receiving and sending traffic. You can manage security groups in\\xA0the <a href=\\"https://kz.console.yandex.cloud/link/vpc/security-groups\\" target=\\"_blank\\" rel=\\"noopener\\">Virtual Private Cloud service</a>.</p>\\n<p>Security groups do\\xA0not restrict traffic to\\xA0the Apache Airflow\\u2122 web interface.</p>\\n"
  }
});"
`;

exports[`transformFileWithBabel transform to TECH_LOCALE 1`] = `
"export default {
  normalMessage: {
    "ru": "\\u0440\\u0443\\u0441\\u0441\\u043A\\u0438\\u0439",
    "en": "english",
    "te-ch": "123456"
  }
};"
`;

exports[`transformFileWithBabel with typograf 1`] = `
"import { intl } from './intl';
export const {
  t
} = intl.createMessages({
  normalMessage: {
    "ru": "\\u0440\\u0443\\u0441\\u0441\\u043A\\u0438\\u0439 \\u044F\\u0437\\u044B\\u043A \\u0434\\u043B\\u0438\\u043D\\u043D\\u043E\\u0435 \\u043F\\u0440\\u0435\\u0434\\u043B\\u043E\\u0436\\u0435\\u043D\\u0438\\u0435 pnpm \\u0441\\u043E\\u043C\\u043D\\u0438\\u0442\\u0435\\u043B\\u044C\\u043D\\u043E, \\u043D\\u043E\\xA0\\u043E\\u043A\\u0435\\u0439",
    "en": "pnpm is\\xA0good package manager"
  }
});"
`;

exports[`transformFileWithBabel without typograf 1`] = `
"import { intl } from './intl';
export const {
  t
} = intl.createMessages({
  normalMessage: {
    "ru": "\\u0440\\u0443\\u0441\\u0441\\u043A\\u0438\\u0439 \\u044F\\u0437\\u044B\\u043A \\u0434\\u043B\\u0438\\u043D\\u043D\\u043E\\u0435 \\u043F\\u0440\\u0435\\u0434\\u043B\\u043E\\u0436\\u0435\\u043D\\u0438\\u0435 pnpm \\u0441\\u043E\\u043C\\u043D\\u0438\\u0442\\u0435\\u043B\\u044C\\u043D\\u043E \\u043D\\u043E \\u043E\\u043A\\u0435\\u0439",
    "en": "pnpm is good package manager"
  }
});"
`;
